ai_graph_color/algorithms/backtracking.py

```diff
@@ -0,0 +1,209 @@
+def run(problem, setup, *params):
+    """
+    This will execute the algorithm
+
+    :param setup: the setup for this algorithm
+    :type setup: :class: ~ai_graph_color.Setup
+    :param problem: the graph to color
+    :type problem: the graph to color
+    :param params: the params for this algorithm
+    :type params: tuple
+    :rtype: list[int] the list of coloring
+    :return: the list of colorings for this graph
+    """
+    yield backtracking_search(problem, params[0], setup)
+
+
+def backtracking_search(graph, num_colors, setup=None):
+    """
+    Searches the graph using backtracking
+
+    :param setup: Setup that contains a logger and a counter
+    :type setup: Setup
+    :param graph: The graph to color
+    :type graph: list[list[int]]
+    :param num_colors: number of colors to try and color with
+    :type num_colors: int
+    :rtype: dict
+    :return: the colored graph
+    """
+    coloring, avail_colors, stack = init(graph, num_colors)
+
+    while True:
+        if len(stack) == 0 or complete(coloring, graph):
+            if setup:
+                setup.logger.debug(
+                    "Finished, final coloring:{}".format(coloring)
+                )
+            yield coloring
+        cur_node = stack[len(stack)-1][1]
+        coloring[cur_node] = stack[len(stack)-1][3]
+        if setup is not None:
+            setup.logger.debug(
+                'Coloring node:{} color{}'.format(
+                    cur_node, stack[len(stack) - 1][3]
+                )
+            )
+            if setup.counter.imcrement():
+                setup.logger.debug(
+                    "Didn't finish, final coloring:{}".format(coloring)
```

Comment:
When setup.counter.increment is True, it doesn't mean you didn't finish. It just means you're being told to preempt yourself.